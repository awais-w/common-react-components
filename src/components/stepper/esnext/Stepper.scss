@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-stepper {
  @include font-standard;
  display: grid;
  height: 100%;
  grid-template-rows: 1fr $super-large;

  &__nav-container {
    position: relative;
    width: 300px;
    transition: width 0.5s;
  }

  &-container {
    &__nav-container {
      height: calc(100vh - 90px);
      padding-top: $super-double;
      overflow-y: auto;
    }
    &__nav-toggle {
      display: flex;
      align-items: center;
      justify-content: center;
      background: $primary-color;
      color: $color-white;
      background: #3777bc;
      color: #fff;
      height: $super-double;
      position: absolute;
      width: 100%;
      top: 0;
      cursor: pointer;
      user-select: none;
      white-space: nowrap;
      z-index: 2;
      i {
        margin-left: $half;
      }
    }
    &__nav-orientation-toggle {
      position: absolute;
      top: 0;
      width: $super-double;
      display: flex;
      height: $super-double;
      align-items: center;
      justify-content: center;
      border: 1px solid rgba(225, 225, 225, 0.2);
      cursor: pointer;
      z-index: 2;
      .cmc-icon-label-value--inline {
        margin: 0;
        opacity: 0.8;
        transition: opacity 0.5s;
        &:hover {
          opacity: 1;
        }
      }
      &-bottom {
        display: flex;
        position: absolute;
        bottom: 0;
        width: 100%;
        justify-content: center;
        align-items: center;
        height: $input-height-regular;
        cursor: pointer;
        color: $primary-color;
        .cmc-icon-label-value {
          margin-right: 0;
          opacity: 0.8;
          transition: opacity 0.5s;
          &:hover {
            opacity: 1;
          }
        }
      }
    }
  }

  &--nav-left {
    grid-template-columns: auto 1fr;
    grid-template-areas: 'navArea contentArea';
  }

  &--nav-right {
    grid-template-columns: 1fr auto;
    grid-template-areas: 'contentArea navArea';
    .cmc-stepper-container__nav-orientation-toggle {
      right: 0;
    }
  }

  &__nav-container {
    grid-area: navArea;
    grid-row: span 2;
    color: $color-grey-charcoal;
    box-shadow: 0 0 $large rgba(0, 0, 0, 0.2);
    z-index: 3;
  }

  &__content-container {
    grid-area: contentArea;
    // Content should always come first
    grid-row: 1;
    overflow-y: auto;
    position: relative;
  }

  &__control-buttons-container {
    grid-area: contentArea;
    grid-row: 2;
    display: grid;
    align-items: center;
    background: $color-alert-teal;
    position: relative;
    z-index: 1;
  }

  .disabled {
    cursor: not-allowed;
    opacity: 0.35;
    filter: alpha(opacity=35);
    box-shadow: none;
  }

  &--nav-minimize {
    .cmc-stepper {
      &__nav-container {
        width: $super-large;
        padding-bottom: $input-height-regular;
      }
      &-container {
        &__nav-toggle {
          i {
            margin: 0;
          }
        }
        &__nav-orientation-toggle -bottom .cmc-icon-label-value {
          margin-right: 0;
        }
      }
    }
  }
}

@include touch-device-bp1024 {
  .cmc-stepper-container {
    &__nav-container {
      padding-top: $input-height-regular;
    }
    &__nav-toggle {
      height: $input-height-regular;
    }
    &__nav-orientation-toggle {
      width: $input-height-regular;
      height: $input-height-regular;
    }
  }
}
