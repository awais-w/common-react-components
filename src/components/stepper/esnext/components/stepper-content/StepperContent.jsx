import React from 'react';
import PropTypes from 'prop-types';

const getStageType = (index, activeStepIndex) => {
  if (index < activeStepIndex) {
    return 'previous';
  }
  if (index === activeStepIndex) {
    return 'current';
  }
  return 'later';
};

const StepperContent = ({
  children,
  activeStepIndex,
  shouldNavBarUpdateOnScroll,
  latestStepReachedIndex,
  onClickPreviousStep,
  onWaypointEnter,
  renderUpToLatestStageReached,
}) =>
  React.Children.map(children, (child, index) => {
    if (index <= (renderUpToLatestStageReached ? latestStepReachedIndex : activeStepIndex)) {
      return React.cloneElement(child, {
        key: `step-${index}`,
        stageType: getStageType(index, activeStepIndex),
        shouldNavBarUpdateOnScroll,
        stepIndex: index,
        activeStepIndex,
        onClickPreviousStep,
        onWaypointEnter,
      });
    }
    return null;
  });

StepperContent.propTypes = {
  /**
   * Set the current step ( Zero based index )
   */
  activeStepIndex: PropTypes.number.isRequired,
  /**
   * The latest stage we've reached up to
   */
  latestStepReachedIndex: PropTypes.number.isRequired,
  /**
   * One or more `<StepperStage/>` components.
   */
  children: PropTypes.node.isRequired,
  /**
   * Determines whether we render a waypoint for each stage so we can update NavBar on scrolling over them
   */
  shouldNavBarUpdateOnScroll: PropTypes.bool.isRequired,
  /**
   * Used to update our parent with a new activeStepID when we click on previous stages
   */
  onClickPreviousStep: PropTypes.func.isRequired,
  /**
   * When we scroll this stage into view, this parent function will be called to update NavBar
   */
  onWaypointEnter: PropTypes.func.isRequired,
  /**
   * Determines whether we should render just to activeIndex or latestStepReachedIndex
   */
  renderUpToLatestStageReached: PropTypes.bool.isRequired,
};

export default StepperContent;
