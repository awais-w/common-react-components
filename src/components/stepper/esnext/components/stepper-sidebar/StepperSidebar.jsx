import React from 'react';
import PropTypes from 'prop-types';
import bem from '@argos/utils/esnext/bem';
import { TransitionGroup } from 'react-transition-group';

// TODO - In the near future we should be importing these directly from node_modules
import IconLabelValue from '@cmc/icon-label-value/esnext';
import NavItem from './components/nav-bar/components/nav-item';
import NavBar from './components/nav-bar';

const cn = bem({ prefix: 'cmc-', block: 'stepper-container' });

const StepperSidebar = ({
  children,
  activeStepIndex,
  currentlyViewingStepIndex,
  onClickPreviousStep,
  onNavToggle,
  onOrientationToggle,
  minimizeNav,
  summaryStyle,
  allowNavigation,
  orientation,
  lastItemDoneOnReach,
}) => (
  <TransitionGroup className={cn('nav-container', null)}>
    <a className={cn('nav-toggle')} onClick={() => onNavToggle(!minimizeNav)}>
      {minimizeNav ? (
        <i className='material-icons'>swap_horiz</i>
      ) : (
        <React.Fragment>
          Minimize this column <i className='material-icons'>compare_arrows</i>
        </React.Fragment>
      )}
    </a>
    <span
      className={cn(`nav-orientation-toggle${minimizeNav ? '-bottom' : ''}`)}
      onClick={() => onOrientationToggle(orientation === 'left-side' ? 'right-side' : 'left-side')}>
      <IconLabelValue
        icon={`icon-to-${orientation === 'left-side' ? 'right' : 'left'}`}
        isCustom
        isInline
        color={minimizeNav ? 'blue' : 'white'}
        tooltipValue={`Move this column to ${orientation === 'left-side' ? 'right' : 'left'}`}
      />
    </span>
    <NavBar activeStepIndex={currentlyViewingStepIndex}>
      {React.Children.map(children, (child, currIndex) => {
        const { icon, isCustomIcon, title, subTitle, summary, shortSummary } = child.props;
        return (
          <NavItem
            key={`navitem-${title}`}
            isComplete={currIndex < activeStepIndex || (lastItemDoneOnReach && children.length === activeStepIndex + 1)}
            isActive={currIndex === activeStepIndex}
            icon={icon}
            isCustomIcon={isCustomIcon}
            title={title}
            subTitle={subTitle || null}
            summary={summary}
            shortSummary={shortSummary}
            summaryStyle={summaryStyle}
            currIndex={currIndex}
            allowNavigation={allowNavigation}
            minimizeNav={minimizeNav}
            handleClick={(e) => {
              if (minimizeNav) onNavToggle(false);
              if (allowNavigation && currIndex < activeStepIndex && !minimizeNav) {
                onClickPreviousStep(e);
              }
            }}
          />
        );
      })}
    </NavBar>
  </TransitionGroup>
);

StepperSidebar.propTypes = {
  /**
   * Set the current step ( Zero based index )
   */
  activeStepIndex: PropTypes.number.isRequired,
  /**
   * What stage we're curent viewing ( Zero based index ) - Allows our nav bar to show which content
   * we're currently viewing ( E.g when scrolling to a previous stage )
   */
  currentlyViewingStepIndex: PropTypes.number.isRequired,
  /**
   * One or more `<StepperStage/>` components.
   */
  children: PropTypes.node.isRequired,
  /**
   * Is Nav minimized
   */
  minimizeNav: PropTypes.bool,
  /**
   * Should the last step be marked as done when reached? (useful in case the last item step is confirmation)
   */
  lastItemDoneOnReach: PropTypes.bool,

  /**
   * Provide the ability to update activeStepIndex in the parent function to a certain index
   */
  onClickPreviousStep: PropTypes.func.isRequired,
  /**
   * Provide the ability to update minimizeNav in the parent function
   */
  onNavToggle: PropTypes.func.isRequired,
  /**
   * Provide the ability to update onOrientationToggle in the parent function
   */
  onOrientationToggle: PropTypes.func.isRequired,
  /**
   * When the summary should be rendered
   */
  summaryStyle: PropTypes.string,
  /**
   * Flag to indicate whether navigation is allowed from stepper stage or not.
   */
  allowNavigation: PropTypes.bool,
};

StepperSidebar.defaultProps = {
  lastItemDoneOnReach: false,
};

export default StepperSidebar;
