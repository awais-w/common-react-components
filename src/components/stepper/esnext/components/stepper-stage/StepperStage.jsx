/**
 * Our <Stepper/> component handles drawing content / headers
 * This just provides a cleaner interface to construct <Stepper/>
 */
import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import Waypoint from 'react-waypoint';
import { Element } from 'react-scroll';
import uuidv4 from 'uuid/v4';
import bem from '@argos/utils/esnext/bem';
import classNames from 'classnames';
import scrollTo from '../../helpers/StepperFunctions';
import './StepperStage.scss';

const cn = bem({ prefix: 'cmc-', block: 'stepper-stage' });

class StepperStage extends PureComponent {
  uniqueId;

  constructor(props) {
    super(props);
    this.uniqueId = uuidv4();
  }

  scrollToMe() {
    scrollTo({
      componentName: this.uniqueId,
      duration: this.props.allowNavigation ? 500 : 0,
      parentID: this.props.parentScrollContainer,
    });
  }

  componentDidMount() {
    this.scrollToMe();
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.activeStepIndex !== this.props.activeStepIndex &&
      this.props.activeStepIndex === this.props.stepIndex
    ) {
      this.scrollToMe();
    }
  }

  render() {
    const {
      // Provided as part of <StepperStage/> props
      title,
      subTitle,
      content,
      // Added by <Stepper />
      stageType,
      stepIndex,
      onWaypointEnter,
      shouldNavBarUpdateOnScroll,
      shouldShowStageTitle,
      allowNavigation,
      onClickPreviousStep,
    } = this.props;

    const classes = classNames(cn(), {
      disabled: stageType !== 'current',
      locked: !allowNavigation,
    });

    return (
      <Fragment>
        <div
          key={`step-${stepIndex}`}
          onClick={
            allowNavigation && stageType === 'previous' ? () => onClickPreviousStep({ desiredIndex: stepIndex }) : null
          }
          className={classes}
          name={this.uniqueId}>
          <Element key={`step-${stepIndex}`} name={`step-${stepIndex}`}>
            {/* TODO - Provide prop */}
            {shouldShowStageTitle && (
              <div className={cn('headings')}>
                <h1>{title}</h1>
                {subTitle && <h4> - {subTitle}</h4>}
              </div>
            )}
            <div className={cn('content')}>{content}</div>
          </Element>
        </div>

        {shouldNavBarUpdateOnScroll && (
          <Waypoint
            onEnter={() => {
              onWaypointEnter({ stepIndex });
            }}
          />
        )}
      </Fragment>
    );
  }
}

StepperStage.propTypes = {
  /**
   * Primary title for this step
   */
  title: PropTypes.string.isRequired,
  /**
   * Show whats been completed in a previous step
   */
  subTitle: PropTypes.string,
  /**
   * Content for the current stage being rendered
   */
  content: PropTypes.node.isRequired,
  /**
   * Set the current step ( Zero based index )
   */
  activeStepIndex: PropTypes.number,
  /**
   * Where abouts this step is located in relation to current ( So we can for example )
   * Provide the ability to navigate back to stages marked as "Previous"
   */
  stageType: PropTypes.oneOf(['current', 'previous', 'later']),
  /**
   * The index of this stage in relation to other <StepperStage />'s
   */
  stepIndex: PropTypes.number,
  /**
   * When we scroll this stage into view, this parent function will be called to update NavBar
   */
  onWaypointEnter: PropTypes.func,
  /**
   * Used to update our parent with a new activeStepID when we click on previous stages
   */
  onClickPreviousStep: PropTypes.func,
  /**
   * Helper function in <Stepper/> to scroll to a component given it's container ID
   */
  shouldNavBarUpdateOnScroll: PropTypes.bool,
  /**
   * Should we show a title above this current stage content
   */
  shouldShowStageTitle: PropTypes.bool,
  /**
   * The parent container within which the scrolling needs to happen.
   */
  parentScrollContainer: PropTypes.string,
  /**
   * Flag to indicate whether navigation is allowed from stepper stage or not.
   */
  allowNavigation: PropTypes.bool,
};

StepperStage.defaultProps = {
  subTitle: '',
  // provide default values for those provided by <Stepper/>
  activeStepIndex: 0,
  stepIndex: 0,
  stageType: 'current',
  onWaypointEnter: () => {},
  onClickPreviousStep: () => {},
  scrollTo: () => {},
  shouldNavBarUpdateOnScroll: true,
  shouldShowStageTitle: false,
  parentScrollContainer: '',
};

export default StepperStage;
