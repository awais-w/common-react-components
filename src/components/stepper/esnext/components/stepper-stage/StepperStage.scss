@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-stepper-stage {
  @include font-standard;
  min-height: 100%;
  padding: $double;
  &__headings {
    display: inline-block;

    h1,
    h4 {
      display: inherit;
    }
  }
  &__page-divider {
    margin-bottom: $large;
  }
}
