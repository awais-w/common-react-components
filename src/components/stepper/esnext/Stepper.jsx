import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import uuidv4 from 'uuid/v4';
import bem from '@argos/utils/esnext/bem';

import isEmpty from '@cmc/utils/esnext/isEmpty';
import StepperStage from './components/stepper-stage';
import StepperSidebar from './components/stepper-sidebar';
import StepperContent from './components/stepper-content';
import './Stepper.scss';

const cn = bem({ prefix: 'cmc-', block: 'stepper' });
export default class Stepper extends PureComponent {
  constructor(props) {
    super(props);
    /**
     * "currentlyViewingStepIndex" - Manages what we're scrolling over to update NavBar
     * "latestStepReachedIndex" - Allow us to draw upto the latest step we've reached step we've
     *  reached even after we've navigated back
     */
    this.state = {
      activeStepIndex: props.activeStepIndex,
      currentlyViewingStepIndex: 0,
      latestStepReachedIndex: 0,
      parentScrollContainer: uuidv4(),
      minimizeNav: false,
      navOrientation: this.props.orientation,
    };

    this.handleClickPreviousStep = this.handleClickPreviousStep.bind(this);
    this.handleWaypointEnter = this.handleWaypointEnter.bind(this);
    this.handleNavToggle = this.handleNavToggle.bind(this);
    this.handleNavOrientation = this.handleNavOrientation.bind(this);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const stateChanges = { activeStepIndex: nextProps.activeStepIndex };
    if (nextProps.activeStepIndex !== prevState.activeStepIndex) {
      // If we've clicked back / forward in our parent our NavBar should reflect that now
      stateChanges.currentlyViewingStepIndex = nextProps.activeStepIndex;
      if (nextProps.activeStepIndex > prevState.activeStepIndex) {
        /**
         * If we've moved to a later step Set latestStepReachedIndex to new step,
         * so if we navigate back in the future we'll still draw up to this
         */
        stateChanges.latestStepReachedIndex = nextProps.activeStepIndex;
      }
    }
    return stateChanges;
  }

  /**
   * provided an index update our internal state + parent state
   * @param {number} desiredIndex - Define a step index to navigate to
   */
  handleClickPreviousStep({ desiredIndex }) {
    // This is internal state used to correcty highlight <NavItem/> when scrolling
    this.setState(() => ({
      currentlyViewingStepIndex: desiredIndex,
    }));

    this.props.handleStep({ desiredIndex });
  }

  /**
   * Maintain the correct step showing on the left when we scroll back / forward through steps
   * @param {number} stepIndex - The step our <NavBar/> should be showing
   */
  handleWaypointEnter({ stepIndex }) {
    this.setState({
      currentlyViewingStepIndex: stepIndex,
    });
  }

  handleNavToggle(toggler) {
    this.setState({
      minimizeNav: toggler,
    });
  }

  handleNavOrientation(position) {
    this.setState({
      navOrientation: position,
    });
  }

  componentDidUpdate(oldProps) {
    const newProps = this.props;
    if (oldProps.orientation !== newProps.orientation) {
      this.setState({ navOrientation: newProps.orientation });
    }
  }

  render() {
    const {
      children,
      activeStepIndex,
      shouldNavBarUpdateOnScroll,
      renderUpToLatestStageReached,
      summaryStyle,
      lastItemDoneOnReach,
    } = this.props;
    const {
      currentlyViewingStepIndex,
      latestStepReachedIndex,
      parentScrollContainer,
      minimizeNav,
      navOrientation,
    } = this.state;

    // Check for isEmpty, because for any reason if any stage does not contain allowNavigation flag it should allow navigation.
    const allowNavigation =
      isEmpty(children[activeStepIndex].props.allowNavigation) || children[activeStepIndex].props.allowNavigation;

    return (
      <div
        style={{ height: this.props.height }}
        className={cn(null, {
          'nav-left': navOrientation === 'left-side',
          'nav-right': navOrientation === 'right-side',
          'nav-minimize': minimizeNav,
        })}>
        <div className={cn('nav-container')}>
          <StepperSidebar
            orientation={navOrientation}
            minimizeNav={minimizeNav}
            activeStepIndex={activeStepIndex}
            currentlyViewingStepIndex={currentlyViewingStepIndex}
            onClickPreviousStep={this.handleClickPreviousStep}
            onNavToggle={this.handleNavToggle}
            onOrientationToggle={this.handleNavOrientation}
            summaryStyle={summaryStyle}
            allowNavigation={allowNavigation}
            lastItemDoneOnReach={lastItemDoneOnReach}>
            {/* Render each <StepperStage/> title / subtitle / summary  */}
            {children}
          </StepperSidebar>
        </div>
        <div className={cn('content-container')} id={parentScrollContainer}>
          <StepperContent
            shouldNavBarUpdateOnScroll={shouldNavBarUpdateOnScroll}
            latestStepReachedIndex={latestStepReachedIndex}
            activeStepIndex={activeStepIndex}
            onClickPreviousStep={this.handleClickPreviousStep}
            onWaypointEnter={this.handleWaypointEnter}
            renderUpToLatestStageReached={renderUpToLatestStageReached}>
            {/* Render each <StepperStage/> content + navigation  / subtitle / summary  */}
            {React.Children.map(children, (child) =>
              React.cloneElement(child, {
                parentScrollContainer: this.state.parentScrollContainer,
                allowNavigation,
              }),
            )}
          </StepperContent>
        </div>
        {/* Display stages button @ bottom of stepper */}
        <div className={cn('control-buttons-container')}>{children[activeStepIndex].props.transitionButton}</div>
      </div>
    );
  }
}

Stepper.propTypes = {
  /**
   * Location of the <NavBar />
   */
  navOrientation: PropTypes.oneOf(['left-side', 'right-side']),
  /**
   * Set the current step ( Zero based index )
   */
  activeStepIndex: PropTypes.number.isRequired,
  /**
   * One or more `<StepperStage/>` components.
   */
  children: PropTypes.oneOfType([
    PropTypes.shape({
      type: PropTypes.oneOf([StepperStage]),
    }),
    PropTypes.arrayOf(
      PropTypes.shape({
        type: PropTypes.oneOf([StepperStage]),
      }),
    ),
  ]).isRequired,
  /**
   * Provide the ability to update activeStepIndex in the parent function to a certain index
   */
  handleStep: PropTypes.func.isRequired,
  /**
   * A set PX height can be provided if needed
   */
  height: PropTypes.string,
  /**
   * Determines whether we render a waypoint for each stage so we can update NavBar on scrolling over them
   */
  shouldNavBarUpdateOnScroll: PropTypes.bool,
  /**
   * Determines whether we should render just to activeIndex or latestStepReachedIndex
   */
  renderUpToLatestStageReached: PropTypes.bool, // TODO: migrate to common-react-components
  /**
   * When the summary should be rendered
   */
  summaryStyle: PropTypes.string,
};

Stepper.defaultProps = {
  orientation: 'left-side',
  height: '100%',
  shouldNavBarUpdateOnScroll: true,
};
