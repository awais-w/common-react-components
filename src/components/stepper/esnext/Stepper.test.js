import React from 'react';
import { shallow } from 'enzyme';
import Stepper from './Stepper';
import StepperStage from './components/stepper-stage';
import StepperSidebar from './components/stepper-sidebar';
import StepperContent from './components/stepper-content';

const defaultProps = {
  orientation: 'left-side',
  height: '100%',
  shouldNavBarUpdateOnScroll: true,
  activeStepIndex: 0,
  minimizeNav: false,
  navOrientation: 'left-side',
  handleStep: jest.fn(),
};

const nextProps = {
  orientation: 'left-side',
  height: '50%',
  shouldNavBarUpdateOnScroll: true,
  activeStepIndex: 1,
};

// Mock the uuid used in StepperStage
jest.mock('uuid/v4', () => () => '00000000-0000-0000-0000-000000000000');

const setup = ({ props }) => {
  const component = shallow(
    <Stepper {...props}>
      <StepperStage allowNavigation={true} />
      <StepperStage />
      <StepperStage allowNavigation={false} />
    </Stepper>,
  );
  return {
    wrapper: component,
    getSidebar: () => component.find(StepperSidebar),
    getSidebarStage: (index) => component.find(StepperSidebar).find(StepperStage).at(index),
    getContent: () => component.find(StepperContent),
    getContentStage: (index) => component.find(StepperContent).find(StepperStage).at(index),
    getCurrentlyViewingStepIndex: () => component.state('currentlyViewingStepIndex'),
    getLatestStepReachedIndex: () => component.state('latestStepReachedIndex'),
    getMinimizeNavState: () => component.state('minimizeNav'),
    getNavOrientationState: () => component.state('navOrientation'),
  };
};

describe('Stepper', () => {
  describe('base tests', () => {
    const component = setup({ props: defaultProps });
    it('exists', () => {
      expect(component.wrapper).toExist();
    });
    it('should match default snapshot', () => {
      expect(component.wrapper).toMatchSnapshot();
    });
    it('should render one sidebar and one content', () => {
      expect(component.getSidebar()).toHaveLength(1);
      expect(component.getContent()).toHaveLength(1);
    });
  });

  describe('when componentWillReceiveProps is called', () => {
    describe('with a different activeStepIndex', () => {
      const component = setup({ props: defaultProps });
      component.wrapper.setProps(nextProps);
      it('should update currentlyViewingStepIndex to the activeStepIndex', () => {
        expect(component.getCurrentlyViewingStepIndex()).toEqual(nextProps.activeStepIndex);
      });
      describe('should update activeStepIndex accordingly', () => {
        it('should update latestStepReachedIndex if the new is greater than old', () => {
          expect(component.getLatestStepReachedIndex()).toEqual(nextProps.activeStepIndex);
        });
        it('should not revert latestStepIndexReached if we revert the activeStepIndex change', () => {
          component.wrapper.setProps({ activeStepIndex: 0 });
          expect(component.getLatestStepReachedIndex()).toEqual(nextProps.activeStepIndex);
        });
      });
    });

    describe('with the same activeStepIndex', () => {
      const component = setup({ props: defaultProps });
      component.wrapper.setProps(defaultProps);
      it('does not update anything', () => {
        expect(component.getCurrentlyViewingStepIndex()).toEqual(defaultProps.activeStepIndex);
        expect(component.getLatestStepReachedIndex()).toEqual(defaultProps.activeStepIndex);
      });
    });
  });

  describe('when handle click previous step is called', () => {
    const component = setup({ props: defaultProps });
    const index = 2;
    component.wrapper.instance().handleClickPreviousStep({ desiredIndex: index });
    it('should set currentlyViewingStepIndex to desired index', () => {
      expect(component.getCurrentlyViewingStepIndex()).toEqual(index);
    });
    it('should call handleStep', () => {
      expect(component.wrapper.instance().props.handleStep).toBeCalled();
    });
  });

  describe('when handle waypoint enter is called', () => {
    const component = setup({ props: defaultProps });
    const index = 2;
    component.wrapper.instance().handleWaypointEnter({ stepIndex: index });
    it('should set currentlyViewingStepIndex to desired index', () => {
      expect(component.getCurrentlyViewingStepIndex()).toEqual(index);
    });
  });

  describe('when handle Nav Toggle is called', () => {
    const component = setup({ props: defaultProps });
    component.wrapper.setProps(defaultProps);
    component.wrapper.instance().handleNavToggle(!defaultProps.minimizeNav);
    it('should toggle minimizeNav State', () => {
      expect(component.getMinimizeNavState()).toEqual(true);
    });

    it('has the right class on the container', () =>
      expect(component.wrapper).toHaveClassName('cmc-stepper--nav-minimize'));
  });

  describe('when handle Nav Orientation is called', () => {
    const component = setup({ props: defaultProps });
    component.wrapper.instance().handleNavOrientation('right-side');
    it('should toggle navOrientation State', () => {
      expect(component.getNavOrientationState()).toEqual('right-side');
    });

    it('has the right class on the container', () =>
      expect(component.wrapper).toHaveClassName('cmc-stepper--nav-right'));
  });
});

describe("allowNavigation prop is passed down correctly to <StepperStage>'s", () => {
  const component = setup({ props: defaultProps });
  it('is passed explicitly down for sidebar stages', () => {
    expect(component.getSidebarStage(0).props().allowNavigation).toBe(true);
    expect(component.getSidebarStage(1).props().allowNavigation).toBeUndefined();
    expect(component.getSidebarStage(2).props().allowNavigation).toBe(false);
  });

  describe('allowNavigation value of content stages is equal to value of currentlyViewingIndexStep', () => {
    it('is true when currentlyViewingIndexStep has allowNavigation value of true', () => {
      const viewingIndex = 0;
      component.wrapper.setState({ currentlyViewingStepIndex: viewingIndex });
      const expectedAllowNavigation = component.getContentStage(viewingIndex).props().allowNavigation;
      expect(component.getContentStage(0).props().allowNavigation).toBe(expectedAllowNavigation);
      expect(component.getContentStage(1).props().allowNavigation).toBe(expectedAllowNavigation);
      expect(component.getContentStage(2).props().allowNavigation).toBe(expectedAllowNavigation);
    });
    it('is false when currentlyViewingIndexStep has allowNavigation value of false', () => {
      const viewingIndex = 2;
      component.wrapper.setState({ currentlyViewingStepIndex: viewingIndex });
      const expectedAllowNavigation = component.getContentStage(viewingIndex).props().allowNavigation;
      expect(component.getContentStage(0).props().allowNavigation).toBe(expectedAllowNavigation);
      expect(component.getContentStage(1).props().allowNavigation).toBe(expectedAllowNavigation);
      expect(component.getContentStage(2).props().allowNavigation).toBe(expectedAllowNavigation);
    });
  });
});
