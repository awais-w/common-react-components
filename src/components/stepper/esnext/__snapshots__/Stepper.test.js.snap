// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper base tests should match default snapshot 1`] = `
<div
  className="cmc-stepper cmc-stepper--nav-left"
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="cmc-stepper__nav-container"
  >
    <StepperSidebar
      activeStepIndex={0}
      allowNavigation={true}
      currentlyViewingStepIndex={0}
      lastItemDoneOnReach={false}
      minimizeNav={false}
      onClickPreviousStep={[Function]}
      onNavToggle={[Function]}
      onOrientationToggle={[Function]}
      orientation="left-side"
    >
      <StepperStage
        activeStepIndex={0}
        allowNavigation={true}
        onClickPreviousStep={[Function]}
        onWaypointEnter={[Function]}
        parentScrollContainer=""
        scrollTo={[Function]}
        shouldNavBarUpdateOnScroll={true}
        shouldShowStageTitle={false}
        stageType="current"
        stepIndex={0}
        subTitle=""
      />
      <StepperStage
        activeStepIndex={0}
        onClickPreviousStep={[Function]}
        onWaypointEnter={[Function]}
        parentScrollContainer=""
        scrollTo={[Function]}
        shouldNavBarUpdateOnScroll={true}
        shouldShowStageTitle={false}
        stageType="current"
        stepIndex={0}
        subTitle=""
      />
      <StepperStage
        activeStepIndex={0}
        allowNavigation={false}
        onClickPreviousStep={[Function]}
        onWaypointEnter={[Function]}
        parentScrollContainer=""
        scrollTo={[Function]}
        shouldNavBarUpdateOnScroll={true}
        shouldShowStageTitle={false}
        stageType="current"
        stepIndex={0}
        subTitle=""
      />
    </StepperSidebar>
  </div>
  <div
    className="cmc-stepper__content-container"
    id="00000000-0000-0000-0000-000000000000"
  >
    <StepperContent
      activeStepIndex={0}
      latestStepReachedIndex={0}
      onClickPreviousStep={[Function]}
      onWaypointEnter={[Function]}
      shouldNavBarUpdateOnScroll={true}
    >
      <StepperStage
        activeStepIndex={0}
        allowNavigation={true}
        key=".0"
        onClickPreviousStep={[Function]}
        onWaypointEnter={[Function]}
        parentScrollContainer="00000000-0000-0000-0000-000000000000"
        scrollTo={[Function]}
        shouldNavBarUpdateOnScroll={true}
        shouldShowStageTitle={false}
        stageType="current"
        stepIndex={0}
        subTitle=""
      />
      <StepperStage
        activeStepIndex={0}
        allowNavigation={true}
        key=".1"
        onClickPreviousStep={[Function]}
        onWaypointEnter={[Function]}
        parentScrollContainer="00000000-0000-0000-0000-000000000000"
        scrollTo={[Function]}
        shouldNavBarUpdateOnScroll={true}
        shouldShowStageTitle={false}
        stageType="current"
        stepIndex={0}
        subTitle=""
      />
      <StepperStage
        activeStepIndex={0}
        allowNavigation={true}
        key=".2"
        onClickPreviousStep={[Function]}
        onWaypointEnter={[Function]}
        parentScrollContainer="00000000-0000-0000-0000-000000000000"
        scrollTo={[Function]}
        shouldNavBarUpdateOnScroll={true}
        shouldShowStageTitle={false}
        stageType="current"
        stepIndex={0}
        subTitle=""
      />
    </StepperContent>
  </div>
  <div
    className="cmc-stepper__control-buttons-container"
  />
</div>
`;
